network/p2p/unicast/message_rate_limiter.go

```diff
@@ -0,0 +1,89 @@
+package unicast
+
+import (
+	"time"
+
+	"golang.org/x/time/rate"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+
+	"github.com/onflow/flow-go/network/message"
+)
+
+// MessageRateLimiterImpl unicast rate limiter that limits the amount of streams that can
+// be created per some configured interval. A new stream is created each time a libP2P
+// node sends a direct message.
+type MessageRateLimiterImpl struct {
+	limiters                 *rateLimiterMap
+	limit                    rate.Limit
+	burst                    int
+	now                      GetTimeNow
+	rateLimitLockoutDuration time.Duration // the amount of time that has to pass before a peer is allowed to connect
+}
+
+// NewMessageRateLimiter returns a new MessageRateLimiterImpl. The cleanup loop will be started in a
+// separate goroutine and should be stopped by calling Close.
+func NewMessageRateLimiter(limit rate.Limit, burst, lockoutDuration int, opts ...RateLimiterOpt) *MessageRateLimiterImpl {
```

Comment:
A general comment: the bandwidth and message rate limiter are not covered by their own unit tests, and they only have integration tests in `middleware_test.go` that cover an integration of the full stack of our networking layer together with the limiters. It adds value to also implement unit tests for these individual limiters so that bugs and edge cases can be detected earlier in isolation than through an integration.